R/utils.R

```diff
@@ -288,28 +288,47 @@ check_attributes <- function(attributes, call = caller_env()) {
 }
 
 # apple + spark drive config (#651) --------------------------------------------
-configure_spark <- function(call = caller_env()) {
+# Method will attempt to:
+# 1. Locate the simba driver config using the user supplied callback.  Callback
+# method defaults to locate_config_spark.
+# 2. Inspect the config for some settings that can impact how our package
+# performs.
+# 3. If action == "modify" then we attempt to modify the config in-situ.
+# 4. Otherwise we throw a warning asking the user to revise.
+configure_simba <- function(locate_config_callback = (function() character()),
```

Comment:
```suggestion
configure_simba <- function(locate_config_callback = character(),
```

Poking at this for a bit, it's unclear to me why this is a function rather than a character string? In this function, we just call the inputted function which is inlined as returning a character everywhere it's currently used, and then use that character result as usual. Maybe this function just takes the config file as a string and we pass e.g. `snowflake_simba_config(args$driver)` rather than `function() snowflake_simba_config(args$driver)`?

Reply:
Hey Simon:

Wrote it as a callback because in may cases ( not macos, or  `no_config_override` is set ) it may not get executed at all.  Can move those checks outside of the method, but it felt like they belong within ( and they would need to be duplicated for each call ).

Let me know if you think it's worth trying to chase this down further / refactor more.

Reply:
Gotcha! Since those callbacks take some time (and could even raise some unanticipated error), you opt to wait to evaluate the code until you're sure you need its result.

We're actually lucky here in that [R will evaluate function arguments lazily](https://adv-r.hadley.nz/functions.html#lazy-evaluation). So, if you pass a function call as an argument to `f()`, do some checks, and then reference the result of the function call after the checks, the function call will never actually evaluate if the check returns beforehand. e.g.:

``` r
fn_a <- function() {
  cli::cli_abort("Stoppp!")
}

fn_b <- function(x) {
  if (TRUE) {
    cli::cli_abort("{.fun fn_a} hasn't evaluated yet!")
  }
  
  fn_a
}

fn_b(x = fn_a())
#> Error in `fn_b()`:
#> ! `fn_a()` hasn't evaluated yet!
```

<sup>Created on 2024-11-12 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

...the implication here being that we can pass `snowflake_simba_config(args$driver)` directly as the function argument and it won't actually run until those Not macOS or No `no_config_override` checks have evaluated.